EasyLM/models/llama/llama_train_ppo.py

```diff
@@ -0,0 +1,421 @@
+    rewards = rewards / std
+    return rewards
+
+def reduce_mean(x, mask):
+    return jnp.sum(x * mask) / jnp.sum(mask)
+
+
+def ppo_loss(
+    policy_model, value_model,
+    policy_params, value_params,
+    rng,
+    input_ids, attention_mask, continuation_mask, old_continuations_logps, old_continuations_values, rewards,
+):
+    if FLAGS.whiten_rewards:
+        rewards = whiten(rewards, continuation_mask, shift_mean=False)
+    lastgaelam = 0
+    advantages_reversed = []
+    len = continuation_mask.shape[1]
+    for t in reversed(range(len)):
+        delta = rewards[:, t] + FLAGS.gamma * old_continuations_values[:, t + 1] - old_continuations_values[:, t]
+        lastgaelam = delta + FLAGS.gamma * FLAGS.lam * lastgaelam
+        advantages_reversed.append(lastgaelam)
+    advantages = jnp.stack(advantages_reversed[::-1], axis=1)
+    advantages = advantages * continuation_mask
+    returns = advantages + old_continuations_values
+    advantages = whiten(advantages, continuation_mask, shift_mean=True)
+    advantages = jax.lax.stop_gradient(advantages)
+
+    new_continuations_logits = policy_model(input_ids, attention_mask, params=policy_params).logits
+    new_continuations_logps = convert_logits_to_logps(new_continuations_logits, input_ids, continuation_mask) # (B, L)
+    ratio = jnp.exp(new_continuations_logps - old_continuations_logps)
+    ratio = jnp.clip(ratio, 1.0 - FLAGS.cliprange, 1.0 + FLAGS.cliprange)
+    policy_losses = - advantages * ratio # (B, L)
+    policy_loss = reduce_mean(policy_losses, continuation_mask)
+
+    new_continuations_value = value_model(input_ids, attention_mask, params=value_params).logits[:, :, 0]
+
+    new_continuations_values = jnp.clip(new_continuations_value, old_continuations_values - FLAGS.cliprange_value, old_continuations_values + FLAGS.cliprange_value)
+    value_losses = 0.5 * jnp.square(new_continuations_values - returns)
+    value_loss = reduce_mean(value_losses, continuation_mask)
+
+    loss = FLAGS.policy_loss_coef * policy_loss + FLAGS.value_loss_coef * value_loss
+    return loss
+
+def ppo_step(
+    policy_train_state, reference_train_state, value_train_state, reward_train_state,
+    policy_model, reference_model, value_model, reward_model,
+    rng,
+    batch,
+):
+    '''
+    batch: `prompt_input_ids` (B, PL), `prompt_attention_mask` (B, PL)
+    '''
+
+    # rollout from current policy, output `input_ids` (B, L), `attention_mask` (B, L), `continuation_mask` (B, L)
+    # input_ids, attention_mask, continuation_mask = batch['prompt_input_ids'], batch['prompt_attn_mask'], batch['prompt_attn_mask']
+    generation_config = GenerationConfig(
+        max_new_tokens=32,
+        do_sample=True,
+        temperature=1.0,
+    )
+    pad_token_id = 0
+    outputs = policy_model.generate(
+        input_ids=batch['prompt_input_ids'],
+        generation_config=generation_config,
+        params=policy_train_state.params,
```

Comment:
Cool. What I'm worried about is `policy_model` seems to keep an internal `._params` variable. Ofc this shall be overridden by the `params=` arg we pass into `generate()`. What I don't know how to do is to avoid passing in that internal variable into the jit function (when we pass in `policy_model`).